opensrp-fp/src/main/java/org/smartregister/fp/common/task/FinalizeVisitFormTask.java

```diff
@@ -0,0 +1,174 @@
+    private final FPFormUtils fpFormUtils;
+    private final ProgressDialog progressDialog;
+    private HashMap<String, String> clientProfileDetail;
+
+    public FinalizeVisitFormTask(Context context, String baseEntityId, int contactNo, Contact contact, String jsonCurrentState) {
+        this.context = context;
+        this.baseEntityId = baseEntityId;
+        this.contactNo = contactNo;
+        this.contact = contact;
+        this.jsonCurrentState = jsonCurrentState;
+        fpFormUtils = new FPFormUtils();
+
+        progressDialog = new ProgressDialog(context);
+        progressDialog.setCancelable(false);
+        progressDialog.setTitle(context.getString(com.vijay.jsonwizard.R.string.loading));
+        progressDialog.setMessage(context.getString(R.string.finalizing_form_message));
+        progressDialog.show();
+    }
+
+    @Override
+    protected Void doInBackground(Void... voids) {
+
+        contact.setJsonForm(fpFormUtils.addFormDetails(jsonCurrentState));
+        contact.setContactNumber(contactNo);
+        FPFormUtils.persistPartial(baseEntityId, contact);
+
+        /*Map<String, String> clientProfileDetail = PatientRepository.getClientProfileDetails(baseEntityId);
+        ContactInteractor contactInteractor = new ContactInteractor();
+        contactInteractor.finalizeContactForm(clientProfileDetail, this);*/
+
+        clientProfileDetail = PatientRepository.getClientProfileDetails(baseEntityId);
+
+        int gestationAge = getGestationAge(clientProfileDetail);
+        boolean isFirst = TextUtils.equals("1", clientProfileDetail.get(DBConstantsUtils.KeyUtils.NEXT_CONTACT));
+        ContactRule contactRule = new ContactRule(gestationAge, isFirst, baseEntityId);
+
+        List<Integer> integerList = FPLibrary.getInstance().getFPRulesEngineHelper()
+                .getContactVisitSchedule(contactRule, ConstantsUtils.RulesFileUtils.CONTACT_RULES);
+
+        int nextContactVisitWeeks = integerList.get(0);
+
+        LocalDate localDate = new LocalDate(clientProfileDetail.get(DBConstantsUtils.KeyUtils.EDD));
+        String nextContactVisitDate = localDate.minusWeeks(ConstantsUtils.DELIVERY_DATE_WEEKS).plusWeeks(nextContactVisitWeeks).toString();
+        int nextContact = getNextContact(clientProfileDetail);
+        WomanDetail womanDetail = getWomanDetail(baseEntityId, nextContactVisitDate, nextContact);
+
+        PatientRepository.updateContactVisitDetails(womanDetail, true);
+
+        try {
+            processFormFieldKeyValues(baseEntityId, new JSONObject(contact.getJsonForm()), String.valueOf(contactNo));
+        }
+        catch (Exception ex) {
+            Timber.e(ex);
+        }
+
+        return null;
+    }
+
+    @Override
+    protected void onPostExecute(Void aVoid) {
+        super.onPostExecute(aVoid);
+        if (progressDialog != null && progressDialog.isShowing()) {
+            progressDialog.dismiss();
+        }
+        Utils.navigateToProfile(context, clientProfileDetail);
```

Comment:
You could use the generics of the AsyncTask and avoid the need to use a lot of variables in the global scope. i.e
`FinalizeVisitFormTask extends AsyncTask<Void, Void, HashMap<String, String>> ` return a hashmap in the task generic and remove the variable `clientProfileDetail` form global scope. This would help prevent unwanted modifications bt other functions and also the code for `onPostExecute` is idempotent

Reply:
yeah, that's the good point. Done